cloudfunctions/scrape_job_listings/cloudfunction.go

```diff
@@ -0,0 +1,132 @@
+/*
+Perform relatively shallow crawl, using heuristics to classify JobListings and traversal continuations.
+
+There are two colly collectors
+1. Traversing links starting from the hn job board
+2. Collecting HTML bodies and their hashes for further processing
+
+For the most part, the heuristics look a links
+- Yc Jobs
+- Lever
+- Greenhouse
+*/
+package cloudfunction
+
+import (
+	"log"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/gocolly/colly/v2"
+	"github.com/gocolly/colly/v2/debug"
+)
+
+var YcPattern = safeCompile(`.*(ycombinator.com)\/companies.*\/jobs\/`)
+
+// handling https://jobs.lever.co/memfault/730541eb-637f-4d9d-9526-8949432f9a34
+// to check for UUIDs
+var LeverPattern = safeCompile(`.*jobs\.lever\.co\/[a-zA-Z0-9._%+-]+\/[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$`)
+
+// handling https://boards.greenhouse.io/supabase/jobs/4777008004
+var GreenhousePattern = safeCompile(`.*(boards.greenhouse.io)\/.*\/jobs\/`)
+
+var LinkPatterns = []*regexp.Regexp{
+	YcPattern,
+	LeverPattern,
+	GreenhousePattern,
+}
+
+func safeCompile(pattern string) *regexp.Regexp {
+	// Compile the regex pattern
+	regex, err := regexp.Compile(pattern)
+	if err != nil {
+		panic(err)
+	}
+	return regex
+}
+
+func CheckLinkPatterns(absoluteLink string) bool {
+	for _, pattern := range LinkPatterns {
+		if pattern.MatchString(absoluteLink) {
+			return true
+		}
+	}
+	return false
+}
+
+func ShouldContinueCrawl(text, absoluteLink string) bool {
+	searchStr := strings.ToLower(text)
+	// ends with /careers
+	// ends with /jobs
+	if strings.HasSuffix(searchStr, "jobs") || strings.HasSuffix(searchStr, "careers") {
+		return true
+	}
+	// looks like a hn posting
+	if strings.Contains(searchStr, "hiring") {
+		return true
+	}
+	return CheckLinkPatterns(absoluteLink)
+
+}
+
+func ShouldFetchJobListing(text, absoluteLink string) bool {
+	return CheckLinkPatterns(absoluteLink)
+}
+
+func InitCrawler() *colly.Collector {
+	// Instantiate default collector
+	ycCrawler := colly.NewCollector(
+		colly.MaxDepth(3),
+		colly.Async(true),
+	)
+
+	jobPostingFetch := colly.NewCollector(
+		colly.Async(false),
+		colly.Debugger(&debug.LogDebugger{}),
+		colly.DisallowedDomains("account.ycombinator.com"),
+	)
+
+	jobPostingFetch.OnResponse(func(r *colly.Response) {
```

Comment:
@devstein I think it makes sense to send a message to a pubsub topic to allow for async processing. Then we can add more advanced filtering I.e. when was the last time we pulled this URL, duplicate checking, etc. downstream

Reply:
Makes sense to me. Also allows us to have scrape multiple "job sources" in parallel the future